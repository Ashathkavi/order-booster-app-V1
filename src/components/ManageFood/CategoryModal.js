import React from 'react'
import Modal from 'react-modal'
import CategoryList from '../../fixtures/FoodCategoryList'

const CategoryModal = (props) => {
    return(
        <Modal
            isOpen={props.modalIsOpen}
            contentLabel = "Please Select a Category"
            onRequestClose = {props.handleCloseModal}
        >
            <h3>Please Select a Category</h3>
            {
                CategoryList().map((category)=>(
                    <button onClick={() =>{                
                        props.handleCategory(category)
                        props.handleCloseModal()
                    }}>{category}</button>
                ))
            }
            
        </Modal>
)}

export default CategoryModal;


/*

*/