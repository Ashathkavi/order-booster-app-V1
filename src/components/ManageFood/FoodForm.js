import React from 'react'
import CategoryModal from '../ManageFood/CategoryModal'
import moment from 'moment'

export default class FoodForm extends React.Component{
    
    constructor(props){

        super(props)
        
        this.state={
            modalIsOpen:false,
            category: !!props.food? props.food.category  : (!!props.foodlargeAvailable ? props.foodlargeAvailable.category :''),
            fullAvailability:props.food ? props.food.largeAvailability : (!!props.foodlargeAvailable ? true: false),
            foodSize:props.food ? props.food.foodSize : (!!props.foodlargeAvailable ? props.foodlargeAvailable.foodSize : 'regular'),
            name:props.food ? props.food.name :  (!!props.foodlargeAvailable ? props.foodlargeAvailable.name :''),
            description:props.food ? props.food.description : (!!props.foodlargeAvailable ? props.foodlargeAvailable.description :  ''),
            amount: props.food ? props.food.amount.toString() : '',
            createdAt: props.food ? props.food.createdAt : moment().valueOf(),
            error:''
        }
        console.log(this.state.category)
    }
    
    

    onSetSize = (e) => {
        this.setState(() => ({
            foodSize:e.target.value
        }))              
    }

    onAvailabilityChange = (e) => {
        //console.log('e.target.value', 'true' === e.target.value)
        if(e.target.value==='false' && this.state.foodSize === 'full'){
            this.setState(() => ({
                foodSize:'regular'
            }))
        }
            
        this.setState(() => ({
            fullAvailability:(e.target.value === 'true')
        }))
    }

    handleCloseModal = () => {    
        this.setState(()=>({
            modalIsOpen:false
        }))
    }
    handleOpenModal = () => {       
        //console.log('open')
        this.setState(()=>({
            modalIsOpen:true
        }))
    }
    handleCategory = (category) => { 
        //console.log('category', category)
        //document.getElementById('selectedCat').value=category; 
        this.setState(()=>({
            category
        }))
    }
    onNameChange = (e) => {
        this.setState(()=>({
            name:e.target.value
        }))
    }
    onDescriptionChange = (e) => {
        this.setState(()=>({
            description:e.target.value
        }))
    }
    onAmountChange = (e) => {
        this.setState(()=>({
            amount:e.target.value
        }))
        //document.getElementById('textInput').value=e.target.value; 
    }

    onCheckFoodsPresents = (foodName, foodSize) => {
        let presents = false
        this.props.foods.map((food)=>{
            if(food.name.toLowerCase() === foodName.toLowerCase() && food.foodSize === foodSize){
                presents = true
                if(!!this.props.food && (this.props.food.name.toLowerCase() === foodName.toLowerCase() && this.props.food.foodSize === foodSize)){
                    presents = false
                }
            }
        })
        return presents
    }

    onSubmit = (e) => {
        e.preventDefault()
        if(!this.state.name || !this.state.category || !this.state.foodSize ){
            this.setState(()=>({
                error:'The field such as NAME, CATEGORY, FULL_AVAILABILITY and FOODSIZE are mandatoroy to fill'
            }))
        }else if(this.onCheckFoodsPresents(this.state.name, this.state.foodSize)){
            this.setState(()=>({
                error:'The Submitted Food is already Available'
            }))
        }else{
            this.setState(()=>({error:''}))
            this.props.onSubmit({
                name:this.state.name,
                amount:parseInt(this.state.amount),
                createdAt:this.state.createdAt,
                description:this.state.description,
                largeAvailability:this.state.fullAvailability,
                foodSize:this.state.foodSize,
                category:this.state.category
            })
        }
    }



    //this.handleSizeRadio()
    render(){
        console.log(this.props,'this.props')
        console.log(this.state,'this.state')
        // console.log('state.modalIsOpen',this.state.modalIsOpen)
        // console.log('state.category',this.state.category)
        // console.log('state.fullAvailability',this.state.fullAvailability)
        // console.log('state.foodSize',this.state.foodSize)
        // console.log('state.description',this.state.description)
        // console.log('state.name',this.state.name)
        return (
                
                
                <form onSubmit={this.onSubmit} className="form">
                    
                    {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <div>
                        { !this.state.category && <span>  Please Select a Food Category First &nbsp;&nbsp;&nbsp;&nbsp;</span>}
                        <button disabled={!!this.props.foodlargeAvailable} type="button" onClick={this.handleOpenModal}> Select Category</button>
                        &nbsp;&nbsp;
                        <input disabled={!!this.props.foodlargeAvailable}  type='text' id='selectedCat' value={this.state.category} disabled required></input>
                    </div>
                    <input 
                        type="text" 
                        placeholder="Name" 
                        disabled={  !this.state.category || !!this.props.foodlargeAvailable}
                        value={this.state.name}
                        onChange={this.onNameChange}
                        className="text-input"
                    />
                    <div>
                        <input type='text' id='textInput' value={this.state.amount} disabled></input>
                        <input  type='range' min='0' max='2000' disabled={!this.state.category} step='10' value={this.state.amount} onChange={(e)=>{
                            this.onAmountChange(e)
                            //this.updateTextInput(e.target.value)
                        }}/>
                    </div>
                    


                    <div >Is Full postion available &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;

                        <input disabled={!!this.props.foodlargeAvailable} type="radio" value={true} checked={this.state.fullAvailability === true} onChange={this.onAvailabilityChange} /> Yes &nbsp;&nbsp;&nbsp;
                        <input disabled={!!this.props.foodlargeAvailable} type="radio" value={false} checked={this.state.fullAvailability === false} onChange={this.onAvailabilityChange} /> No
                    </div>
                    <div hidden={!this.state.fullAvailability} >Set the size of the potion &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;

                        <input disabled={!!this.props.foodlargeAvailable} type="radio" value='regular' checked={this.state.foodSize === 'regular'} onChange={this.onSetSize} /> Normal &nbsp;&nbsp;&nbsp;
                        <input disabled={!!this.props.foodlargeAvailable} type="radio" value='full' checked={this.state.foodSize === 'full'} onChange={this.onSetSize} /> Full
                    </div>

                    
                                
                    <CategoryModal modalIsOpen={this.state.modalIsOpen} handleCloseModal={this.handleCloseModal} handleCategory={this.handleCategory}/>
                    
                    <textarea 
                        placeholder="Add a description about this food" 
                        disabled={!this.state.category}
                        value={this.state.description}
                        onChange={this.onDescriptionChange}   
                        className="textarea"
                    />
                    
                    <div>
                        <button className="button">Save Food</button>

                    </div>

                </form>
        )
    }
}