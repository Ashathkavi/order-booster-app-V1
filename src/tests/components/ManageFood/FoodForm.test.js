import React from 'react'
import {shallow} from 'enzyme'
import FoodForm from '../../../components/ManageFood/FoodForm'
import sampleFoods from '../../../fixtures/sampleFoods'
import CategoryModal from '../../../components/ManageFood/CategoryModal'


const sample_foods = sampleFoods()

test('should render FoodForm correctly', () => {
    const wrapper = shallow(<FoodForm/>)
    expect(wrapper).toMatchSnapshot()
})

test('should render FoodForm correctly with food date', () => {
    const wrapper = shallow(<FoodForm food= {sample_foods[0]}/>)
    expect(wrapper).toMatchSnapshot()
})


test('should render error for invalid form submission', () => {
    const wrapper = shallow(<FoodForm/>)
    expect(wrapper).toMatchSnapshot()
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    })
    expect(wrapper.state('error').length).toBeGreaterThan(0)
    expect(wrapper).toMatchSnapshot()
})



test('should set size on input change', () => {
    const value = 'regular'
    const wrapper = shallow(<FoodForm/>)
    wrapper.find('input').at(6).simulate('change', {
        target:{value},
        persist: () => { }
    })
    expect(wrapper.state('foodSize')).toBe(value)
})


test('should set availability change on input change', () => {
    const valueFullAvailability = false
    const valueSize = 'regular'
    const wrapper = shallow(<FoodForm/>)
    wrapper.find('input').at(3).simulate('change', {
        target:{valueFullAvailability},
        persist: () => { }
    })
    wrapper.find('input').at(4).simulate('change', {
        target:{valueSize},
        persist: () => { }
    })
    expect(wrapper.state('foodSize')).toBe(valueSize)
    expect(wrapper.state('fullAvailability')).toBe(valueFullAvailability)
})


test('should not set size as full is availability is false ', () => {
    const valueFullAvailability = false
    const valueSizeFull = 'full'
    const valueSizeRegular = 'regular'
    const wrapper = shallow(<FoodForm/>)
    wrapper.find('input').at(3).simulate('change', {
        target:{valueFullAvailability},
        persist: () => { }
    })
    wrapper.find('input').at(4).simulate('change', {
        target:{valueSizeFull},
        persist: () => { }
    })
    
    expect(wrapper.state('fullAvailability')).toBe(valueFullAvailability)
    expect(wrapper.state('foodSize')).toBe(valueSizeRegular)
})


test('should set name on input change', () => {
    const value = 'chicken soup'
    const wrapper = shallow(<FoodForm/>)
    wrapper.find('input').at(1).simulate('change', {
        target:{value},
        persist: () => { }
    })
    expect(wrapper.state('name')).toBe(value)
})

test('should set amount on input change', () => {
    const value = 25
    const wrapper = shallow(<FoodForm/>)
    wrapper.find('input').at(2).simulate('change', {
        target:{value},
        persist: () => { }
    })
    expect(wrapper.state('amount')).toBe(value)
})


test('should set description on textarea change', () => {
    const value = 'more spicy soup'
    const wrapper = shallow(<FoodForm/>)
    wrapper.find('textarea').at(0).simulate('change', {
        target:{value},
        persist: () => { }
    })
    expect(wrapper.state('description')).toBe(value)
})



test('should call onSubmit props for valid form submission', () => {
    const onSubmitSpy = jest.fn()
    const wrapper = shallow(<FoodForm food={sample_foods[0]} onSubmit={onSubmitSpy}/>)
    wrapper.find('form').simulate('submit', {
        preventDefault:() => {}
    })
    expect(wrapper.state('error')).toBe('')
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        name:sample_foods[0].name,
        amount:sample_foods[0].amount,
        createdAt:sample_foods[0].createdAt,
        description:sample_foods[0].description,
        largeAvailability:sample_foods[0].largeAvailability,
        foodSize:sample_foods[0].foodSize,
        category:sample_foods[0].category
    
    })
})


test('should set the modal visibility  when click the category buttton', () => {
    const wrapper = shallow(<FoodForm/>)
    //console.log(wrapper.state('isModalOpen'))
    wrapper.find('button').at(0).simulate('click')
    expect(wrapper.state('modalIsOpen')).toEqual(true)
})


test('should set the modal visibility  when close the modal ', () => {
    const wrapper = shallow(<FoodForm/>)
    wrapper.find(CategoryModal).prop('handleCloseModal')()
    expect(wrapper.state('modalIsOpen')).toEqual(false)
})


test('should set category state ', () => {
    const selectedCategory = 'soup'
    const wrapper = shallow(<FoodForm/>)
    wrapper.find(CategoryModal).prop('handleCategory')(selectedCategory)
    expect(wrapper.state('category')).toEqual(selectedCategory)
})
