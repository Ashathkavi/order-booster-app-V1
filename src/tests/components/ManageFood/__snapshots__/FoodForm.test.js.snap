// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render FoodForm correctly 1`] = `
<div>
  <p>
    Please Select a Food Category First
  </p>
  <form
    onSubmit={[Function]}
  >
    <button
      onClick={[Function]}
      type="button"
    >
       Select Category
    </button>
    <input
      disabled={true}
      id="selectedCat"
      required={true}
      type="text"
      value=""
    />
    <input
      disabled={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      disabled={true}
      max="2000"
      min="0"
      onChange={[Function]}
      step="10"
      type="range"
      value=""
    />
    <input
      disabled={true}
      id="textInput"
      type="text"
      value=""
    />
    <div>
      Is Full postion available::::::
      <input
        checked={false}
        onChange={[Function]}
        type="radio"
        value={true}
      />
       Yes
      <input
        checked={true}
        onChange={[Function]}
        type="radio"
        value={false}
      />
       No
    </div>
    <div
      hidden={true}
    >
      Set the size of the potion:::::
      <input
        checked={true}
        onChange={[Function]}
        type="radio"
        value="regular"
      />
       Normal
      <input
        checked={false}
        onChange={[Function]}
        type="radio"
        value="full"
      />
       Full
    </div>
    <CategoryModal
      handleCategory={[Function]}
      handleCloseModal={[Function]}
      modalIsOpen={false}
    />
    <textarea
      disabled={true}
      onChange={[Function]}
      placeholder="Add a description about this food"
      value=""
    />
    <button>
      Add Food
    </button>
  </form>
</div>
`;

exports[`should render FoodForm correctly with food date 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <button
      onClick={[Function]}
      type="button"
    >
       Select Category
    </button>
    <input
      disabled={true}
      id="selectedCat"
      required={true}
      type="text"
      value="Fried Rice"
    />
    <input
      disabled={false}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Chicken Fried Rice"
    />
    <input
      disabled={false}
      max="2000"
      min="0"
      onChange={[Function]}
      step="10"
      type="range"
      value="350"
    />
    <input
      disabled={true}
      id="textInput"
      type="text"
      value="350"
    />
    <div>
      Is Full postion available::::::
      <input
        checked={true}
        onChange={[Function]}
        type="radio"
        value={true}
      />
       Yes
      <input
        checked={false}
        onChange={[Function]}
        type="radio"
        value={false}
      />
       No
    </div>
    <div
      hidden={false}
    >
      Set the size of the potion:::::
      <input
        checked={true}
        onChange={[Function]}
        type="radio"
        value="regular"
      />
       Normal
      <input
        checked={false}
        onChange={[Function]}
        type="radio"
        value="full"
      />
       Full
    </div>
    <CategoryModal
      handleCategory={[Function]}
      handleCloseModal={[Function]}
      modalIsOpen={false}
    />
    <textarea
      disabled={false}
      onChange={[Function]}
      placeholder="Add a description about this food"
      value=""
    />
    <button>
      Add Food
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 1`] = `
<div>
  <p>
    Please Select a Food Category First
  </p>
  <form
    onSubmit={[Function]}
  >
    <button
      onClick={[Function]}
      type="button"
    >
       Select Category
    </button>
    <input
      disabled={true}
      id="selectedCat"
      required={true}
      type="text"
      value=""
    />
    <input
      disabled={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      disabled={true}
      max="2000"
      min="0"
      onChange={[Function]}
      step="10"
      type="range"
      value=""
    />
    <input
      disabled={true}
      id="textInput"
      type="text"
      value=""
    />
    <div>
      Is Full postion available::::::
      <input
        checked={false}
        onChange={[Function]}
        type="radio"
        value={true}
      />
       Yes
      <input
        checked={true}
        onChange={[Function]}
        type="radio"
        value={false}
      />
       No
    </div>
    <div
      hidden={true}
    >
      Set the size of the potion:::::
      <input
        checked={true}
        onChange={[Function]}
        type="radio"
        value="regular"
      />
       Normal
      <input
        checked={false}
        onChange={[Function]}
        type="radio"
        value="full"
      />
       Full
    </div>
    <CategoryModal
      handleCategory={[Function]}
      handleCloseModal={[Function]}
      modalIsOpen={false}
    />
    <textarea
      disabled={true}
      onChange={[Function]}
      placeholder="Add a description about this food"
      value=""
    />
    <button>
      Add Food
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 2`] = `
<div>
  <p>
    Please Select a Food Category First
  </p>
  <p>
    The field such as NAME, CATEGORY, FULL_AVAILABILITY and FOODSIZE are mandatoroy to fill
  </p>
  <form
    onSubmit={[Function]}
  >
    <button
      onClick={[Function]}
      type="button"
    >
       Select Category
    </button>
    <input
      disabled={true}
      id="selectedCat"
      required={true}
      type="text"
      value=""
    />
    <input
      disabled={true}
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      disabled={true}
      max="2000"
      min="0"
      onChange={[Function]}
      step="10"
      type="range"
      value=""
    />
    <input
      disabled={true}
      id="textInput"
      type="text"
      value=""
    />
    <div>
      Is Full postion available::::::
      <input
        checked={false}
        onChange={[Function]}
        type="radio"
        value={true}
      />
       Yes
      <input
        checked={true}
        onChange={[Function]}
        type="radio"
        value={false}
      />
       No
    </div>
    <div
      hidden={true}
    >
      Set the size of the potion:::::
      <input
        checked={true}
        onChange={[Function]}
        type="radio"
        value="regular"
      />
       Normal
      <input
        checked={false}
        onChange={[Function]}
        type="radio"
        value="full"
      />
       Full
    </div>
    <CategoryModal
      handleCategory={[Function]}
      handleCloseModal={[Function]}
      modalIsOpen={false}
    />
    <textarea
      disabled={true}
      onChange={[Function]}
      placeholder="Add a description about this food"
      value=""
    />
    <button>
      Add Food
    </button>
  </form>
</div>
`;
